xchainkeys.1
.TH XCHAINKEYS 1 "2010-09-16" "Linux" "User manuals"
.SH NAME
\fBxchainkeys\fP \- chained keybindings for X11
.SH SYNOPSIS
\fBxchainkeys\fP [\fIoptions\fP]
.SH DESCRIPTION
xchainkeys is a standalone X11 program to create chained key
bindings similar to those found in the \fBratpoison\fP(1) window
manager or the \fBscreen\fP(1) terminal multiplexer.
.SH OPTIONS
.IP \fB-f\fP,\ \fB--file\fP\ \fI<filename>\fP
Load an alternative configuration file.
.IP \fB-k\fP,\ \fB--keys\fP
Interactively show valid keyspecs for key combinations.
.IP \fB-d\fP,\ \fB--debug\fP
Show debug messages on stdout.
.IP \fB-h\fP,\ \fB--help\fP
Show usage information.
.IP \fB-v\fP,\ \fB--version\fP
Show version information.
.SH CONFIGURATION
The configuration file is located at 

\fI~/.config/xchainkeys/xchainkeys.conf\fP

The \fIXDG_CONFIG_HOME\fP environment variable will be respected in
compliance with the \fBXDG Base Directory Specification\fP (see
\fBREFERENCES\fP below).

An example configuration file has been installed to

\fI__PREFIX__/share/doc/xchainkeys/example.conf\fP
.SS Syntax
Each line may contain either a global setting or a keybinding
definition. Empty lines and lines starting with a numbersign (#) are
ignored.
.SS Global Settings
Global settings are defined as key-value pairs, separated by
whitespace.
.IP \fBtimeout\fP\ \fI<n>\fP\ (default:\ \fB3000\fP)

Wait \fIn\fP milliseconds for another key to be pressed before
aborting an active keychain. A value of 0 will prevent automatic
timeout. This value may be overridden by a \fBtimeout\fP argument
supplied to an \fB:enter\fP action, see \fBActions\fP below.
.IP \fBfeedback\fP\ \fI<\fBon\fP|\fBoff\fP>\fP\ (default:\ \fBon\fP)

Show feedback in a small popup window.
.IP \fBdelay\fP\ \fI<n>\fP\ (default:\ \fB1000\fP)

Show current chain status after \fIn\fP milliseconds of inactivity
in an active chain.
.IP \fBhold\fP\ \fI<n>\fP\ (default:\ \fI<delay>\fP)

Show error messages for \fIn\fP milliseconds after a chain has been
aborted. If not set, the value of \fBdelay\fP will be used.
.IP \fBposition\fP\ \fI<x|offset|\fBmouse\fP|\fBcenter\fP>\ [\fI<y>\fP]\ (default:\ \fBcenter\fP)

The position of the feedback window. An absolute position can be
specified with the \fIx\fP and \fIy\fP arguments. A relative position
can be specified using the offset portion of a standard X geometry
string. The value \fBmouse\fP will cause the window to appear centered
around the current mouse pointer position. The value \fBcenter\fP will
center the window on the default screen.
.IP \fBfont\fP\ \fI<fontstring>\fP\ (default:\ \fBfixed\fP)

The X11 font to use for text in the feedback window.
.IP \fBforeground\fP\ \fI<name|#rrggbb>\fP\ (default:\ \fBblack\fP)

The foreground color of the text and the border of the feedback window.
.IP \fBbackground\fP\ \fI<name|#rrggbb>\fP\ (default:\ \fBwhite\fP)

The background color of the feedback window.
.SS Keybindings
Keybindings are defined using the following syntax:

    \fI<keyspec>\fP \fB[\fP\fI<keyspec>\fP\fB...]\fP :\fI<action>\fP \fB[\fP"\fI<name>\fP"\fB]\fP\ \fB[\fP\fI<arguments>\fP\fB...]\fP

A \fIkeyspec\fP consists of one or more optional \fBmodifier names\fP,
followed by either a symbolic \fBkey name\fP or a numeric keycode,
each separated by a hyphen (-):

    \fI<mod>\fP-\fB[\fP\fI<mod>\fP-\fB...]\fP\fI<keyname|keycode>\fP

Modifier names:

.RS	
    \fIS\fP, \fIShift\fP                    = Shift key
    \fIC\fP, \fICtrl\fP                     = Control key
    \fIA\fP, \fIM\fP, \fIAlt\fP, \fIMeta\fP             = Alt key
    \fIW\fP, \fIH\fP, \fISuper\fP, \fIHyper\fP, \fIWindows\fP = Super key
.RE

The generic modifier names \fIShift\fP, \fILock\fP, \fIControl\fP,
\fIMod1\fP, \fIMod2\fP, \fIMod3\fP, \fIMod4\fP and \fIMod5\fP are
recognized in addition to the above.

Key names correspond to the symbolic key names defined in
\fI/usr/include/X11/keysymdef.h\fP (with their \fIXK_\fP prefix
removed).

Numeric keycodes may be supplied in decimal, hexadecimal or octal
format, e.g. 16 (decimal) = 0x10 (hexadecimal) = 020 octal.

You can use \fBxchainkeys -k\fP to find the proper \fIkeyspec\fP for
key combinations.
.SS Actions

.IP \fB:enter\fP\ [\fBtimeout=\fP\fI<n>\fP]\ [\fBabort=\fP\fI<auto|manual>\fP]

Enters a keychain. The keyboard will be grabbed exclusively until
another keystroke has been processed, or until the chain times out
after \fIn\fP milliseconds. If the key is bound within this chain,
the associated action will be invoked, otherwise the user will be
notified that an unbound key has been pressed.

If the \fPabort\fP argument is set to \fImanual\fP, the chain prefix
will remain activated and further keystrokes will be processed until
the chain is manually aborted with the \fB:abort\fP action. If set to
\fIauto\fP (the default), the chain will abort immediately after the
first keystroke has been processed.

If the \fBtimeout\fP argument is omitted, the global timeout setting
will apply. A value of \fI0\fP disables the timeout for this chain.
.IP \fB:abort\fP
Abort the current keychain, releasing any previous grab on the
keyboard. If not explicitly bound, \fB:abort\fP will be bound to
\fBprefix C-g\fP by default.
.IP \fB:escape\fP
Send the \fBprefix key\fP key of this chain to the currently focused
window and abort the chain. If not explicitly bound, \fB:escape\fP
will be bound to \fBprefix prefix\fP, e.g. for a chain prefixed with
\fBC-t\fP the \fB:escape\fP action will be bound to \fBC-t C-t\fP by
default.

This feature is implemented using \fBXSendEvent(3)\fP. Note that
applications may refuse to accept synthetic events generated with
\fBXSendEvent\fP.
.IP \fB:exec\fP\ [\fBabort=\fP\fIauto\fP]\ \fI<command>\fP
Asynchronously execute \fIcommand\fP as a shell command. If
\fBabort=\fP\fIauto\fP is given, this binding will always abort the
current chain after executing the command, even if the chain was
entered with \fBabort=\fP\fImanual\fP.
.IP \fB:wait\fP\ \fI<command>\fP
Ungrab the keyboard, execute \fIcommand\fP as a shell command, wait
until the \fIcommand\fP has completed and grab the keyboard again if
necessary. Use this action to run commands that also grab the keyboard
(e.g. \fIscrot --select\fP).
.IP \fB:group\fP\ \fB[\fP"\fI<name>\fP"\fB]\fP\ [\fBabort=\fP\fIauto\fP]\ \fI<command>\fP
Associate a binding with a named group and asynchronously execute
\fIcommand\fP as a shell command.

When a \fB:group\fP action is activated, timeout and automatic abort
are disabled for the current chain prefix, and only bindings in the
same group will be recognized for subsequent key presses. Any other
keys will abort the chain and revert the above changes.

If \fBabort=\fP\fIauto\fP is given for particular binding within a
group, this binding will always abort the current chain after
executing the command, effectively overriding the implied
\fBabort=\fP\fImanual\fP option.

In addition, if the key used to abort the chain key is a toplevel
chain \fBprefix key\fP, the corresponding chain will be entered
immediately after this chain aborts. See \fBEXAMPLES\fP below for
possible uses.

If no explicit name is given, the group name "default" is used.

.IP \fB:repeat\fP\ \fI<command>\fP
Deprecated. Equivalent to \fB:group\fP \fB"default"\fP.
.IP \fB:load\fP\ [\fI<filename>\fP]
.IP \fB:reload\fP\ [\fI<filename>\fP]
Clear the current bindings and (re)load the configuration file specified
by \fIfilename\fP. If no file is specified, the current configuration
file is reloaded.
.SH EXAMPLES
Bind the key sequence "Control-t Return" to run xterm:\fP

     C-t Return :exec xterm

\fPThis implies creating a default chain with the prefix key
\fBC-t\fP, containing default bindings for the \fB:abort\fP and
\fB:escape\fP actions, and is thus equivalent to\fB

    C-t :enter timeout=3000 abort=auto
    C-t C-t :escape                   
    C-t C-g :abort                    
    C-t Return :exec xterm            

\fPNote that the number of keys in a keychain is not limited, e.g\fB

     C-t x c h a i n k e y s :exec xmessage "xchainkeys!"

\fPcan be invoked by pressing \fBC-t\fP and then typing "xchainkeys". 

The following example creates a named \fP:group\fP of bindings to
implement a dedicated frame resize mode for the \fBmusca\fP(1) window
manager:\fB

     C-w i :group "resize" musca -c 'resize up'
     C-w k :group "resize" musca -c 'resize down'
     C-w j :group "resize" musca -c 'resize left'
     C-w l :group "resize" musca -c 'resize right'

     C-w ... (other bindings for window management)

\fPInvoke any of the above bindings (e.g. press \fBC-w i\fP) and then
continue to press \fBi\fP, \fBk\fP, \fBj\fP or \fBl\fP to resize the
current musca frame. Since timeout and automatic abort are disabled,
you can continue resizing until you are satisfied with the
result. Then simply press any other key to quit resize mode. If
you want to invoke another window management command right away, you
can use \fBC-w\fP to quit resize mode and immediately enter the
\fBC-w\fP chain (or any other toplevel chain) again.

The benefit of using a named \fB:group\fP over a dedicated chain with
\fBtimeout=0\fP and \fBabort=manual\fP is that you don't have to waste
another prefix key. Instead you can temporarily redefine an existing
chain to contain only a specific subset of keys, behaving in the
desired manner.
.SH KNOWN ISSUES
Some window managers assume that no other program has grabbed any
global key combinations before them, and may fail with a fatal error
when trying to grab a key that has already been grabbed as a prefix
key by xchainkeys. This situation is likely to occur when xchainkeys
is started before the window manager in \fI~/.xinitrc\fP. 

In general, binding the same key combinations in xchainkeys and other
programs should be avoided.
.SH BUGS
This software is currently in beta state. Please file bugs on the
googlecode issue tracker.

http://code.google.com/p/xchainkeys/issues

Use the email address below for questions, feature requests and
comments.
.SH COPYRIGHT
Copyright (C) 2010-2012 Henning Bekel <h.bekel@googlemail.com>
.SH LICENSE
xchainkeys is licensed under the GNU General Public License v3, see
http://www.gnu.org/licenses/gpl.html.
.SH REFERENCES
.IP \fBX11\ key\ symbols\fP
/usr/include/X11/keysymdef.h
.IP \fBXDG\ Base\ Directory\ Specification\fP
http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html
.SH "SEE ALSO"
.BR Xorg (1)
.BR screen (1)
.BR ratpoison (1)
.BR musca (1)


ChangeLog
2012-09-09  Henning Bekel  <h.bekel@googlemail.com>
	* Released version 0.11
	* Added the ability to override manual abort on a per-key basis
	for :exec and :group actions (requested by Salam Taya)

2011-03-28  Henning Bekel  <h.bekel@googlemail.com>

	* Released version 0.10
	* Added new options "position" and "hold" to control the
	positioning of the feedback window and to set the duration of
	error messages being displayed. Based on a patch from Lourens
	Rozema.

2011-03-20  Henning Bekel  <h.bekel@googlemail.com>

	* Released version 0.9.1
	* Fixed a bug in :wait action that left the keyboard grabbed if
	:wait was invoked from a toplevel key

2011-03-12  Henning Bekel  <h.bekel@googlemail.com>

	* Released version 0.9
	* Added new action :wait to execute commands synchronously while
	releasing xchainkey's keyboard grab (requested by tum).

2011-02-11  Henning Bekel  <h.bekel@googlemail.com>

	* Released version 0.8
	* Addded ability to also bind keys by specifying a numeric keycode
	  instead of a keysym

2011-01-28  Henning Bekel  <h.bekel@googlemail.com>

	* Released version 0.7
	* Applied a patch contributed by sssslang that avoids using 100%
	  cpu time while waiting for a timeout to occur.

2010-09-30  Henning Bekel  <h.bekel@googlemail.com>

	* Released version 0.6
	* Added improved feedback for named :group actions
	* Added common modifier synonyms: Ctrl, Alt, Meta, Super, Windows
	* Fixed bogus keys may have been bound on startup from xinit
	* Fixed useless default actions were created when the default keys
	        were already bound to different actions
	* Fixed output streams were not properly flushed
	* Fixed leaked some memory during parsing

2010-09-27  Henning Bekel  <h.bekel@googlemail.com>

	* Released version 0.5
	* Added display of action names in feedback window
	* Added improved error reporting during parsing
	* Fixed spaces in action names were not possible
	* Fixed :abort, :escape, :group were allowed as toplevel bindings
	* Fixed :escape did not imply :abort as documented
	* Fixed parsing failed with unknown words in :enter arguments
	* Code cleanups

2010-09-20  Henning Bekel  <h.bekel@googlemail.com>

	* Released version 0.4
	* Deprecated :repeat action in favor of a new :group action
	* Added :load and :reload actions to load config files
	* Fixed possibility of creating of non-abortable chains
	* Fixed segfault on binding a key sequence multiple times
	* Fixed argument parsing for :enter action
	* Improved error reporting
	* Updated manual page
	* Code cleanups

2010-09-16  Henning Bekel <h.bekel@googlemail.com>

	* Released version 0.3
	* Added timeout and abort options to :enter action
	* Added new action :repeat
	* Added option --file to load an alternative config file
	* Added option feedback [on|off] (requested by tum)
	* Added capitalized generic modifier names (requested by tum)
	* Fixed invalid free()'s in config parsing (reported by tum)
	* Fixed popup window is not raised when shown
	* Updated documentation, added a manual page
	* Code cleanups and minor bugfixes

2010-09-15  Henning Bekel <h.bekel@googlemail.com>

	* Released version 0.2
	* Fixed possibility of invalid free() in xc_show_keys()

2010-09-14  Henning Bekel <h.bekel@googlemail.com>

	* Released version 0.1


src/xchainkeys.c
/* xchainkeys -- chained keybindings for X11
 * Copyright (C) 2010, 2011 Henning Bekel <h.bekel@googlemail.com>
 * Improved timeout code contributed by sssslang.
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.  
 */

#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 500
#endif /* _XOPEN_SOURCE */

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif /* _GNU_SOURCE */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <getopt.h>
#include <sys/time.h>
#include <X11/Xlib.h>

#include "key.h"
#include "binding.h"
#include "popup.h"
#include "util.h"
#include "xchainkeys.h"

XChainKeys_t *xc;

XChainKeys_t* xc_new() {

  XChainKeys_t *self = (XChainKeys_t *) calloc(1, sizeof(XChainKeys_t)); 

  if(NULL == (self->display=XOpenDisplay(NULL))) {
    
    fprintf(stderr, "%s: error: XOpenDisplay() failed for DISPLAY=%s.\n", 
	    PACKAGE_NAME, getenv("DISPLAY")); 
    fflush(stderr);
    
    free(self);
    exit(EXIT_FAILURE);
  }

  XSetErrorHandler(xc_handle_error);

  self->debug = False;
  self->timeout = 3000;
  self->delay = 1000;
  self->hold = -1;
  self->position = "center";
  self->reentry = NULL;
  self->reload = False;

  self->xmodmap = XGetModifierMapping(self->display);
  xc_init_modmask(self);

  self->connection = ConnectionNumber(self->display);

  self->action_names[0] = ":none";
  self->action_names[1] = ":enter";
  self->action_names[2] = ":escape";
  self->action_names[3] = ":abort";
  self->action_names[4] = ":exec";
  self->action_names[5] = ":group";
  self->action_names[6] = ":load";
  self->action_names[7] = ":wait";

  self->root = binding_new();
  self->root->action = XC_ACTION_NONE;
  
  xc_find_config(self);

  return self;
}

void xc_init_modmask(XChainKeys_t *self) {
  unsigned int num, caps, scroll;

  /* setup self->modmask to contain all possible combinations of num,
   * caps and scroll lock modifiers. 
   *
   * Xchainkeys ignores these modifiers by default, e.g. each key will
   * be bound to all possible combinations */

  /* get the keycodes for num, caps and scroll lock */
  num = 
    keycode_to_modifier(self->xmodmap, 
			XKeysymToKeycode(self->display, XStringToKeysym("Num_Lock")));
  caps = 
    keycode_to_modifier(self->xmodmap, 
			XKeysymToKeycode(self->display, XStringToKeysym("Caps_Lock")));
  scroll = 
    keycode_to_modifier(self->xmodmap, 
			XKeysymToKeycode(self->display, XStringToKeysym("Scoll_Lock")));

  /* populate modmask array with all possible combinations */
  self->modmask[0] = 0;
  self->modmask[1] = num;
  self->modmask[2] = caps;
  self->modmask[3] = scroll;
  self->modmask[4] = num | caps;
  self->modmask[5] = num | scroll;
  self->modmask[6] = caps | scroll;
  self->modmask[7] = num | caps | scroll;
}

int xc_handle_error(Display *display, XErrorEvent *event ) {

  /* Xlib error handler. Report the standard error text and exit.
   * Hint if a key is already grabbed by another application */

  char error[1024];
  XGetErrorText( display, event->error_code, error, sizeof(error));
  fprintf(stderr, "error: (X) %s\n", error);
  
  if (event->error_code == BadAccess && event->request_code == 33) {
    fprintf(stderr, "A key is already grabbed by another application " 
	    "(most likely your wm/DE).\n");
  }
  fflush(stderr);   
  exit(EXIT_FAILURE);
}

void xc_show_keys(XChainKeys_t *self) {

  XEvent event;
  Key_t *key;
  KeyCode keycode;
  KeySym keysym;
  char *keystr = "";
  char *keyspec = "";

  if(!xc->debug) {
    version();
    printf("\n");
  }

  printf("Press a key combination to show the corresponding keyspec.\n");
  printf("Press Control-c to quit.\n\n");
  fflush(stdout);

  XGrabKeyboard(self->display, DefaultRootWindow(self->display),
		True, GrabModeAsync, GrabModeAsync, CurrentTime);

  while(True) {
    
    /* block until an event occurs */
    XNextEvent(self->display, &event);
    
    /* we're only interested in KeyPress events */
    if (event.type == KeyPress) {
      
      keycode = ((XKeyPressedEvent*)&event)->keycode;
      
      /* skip modifier key press */
      if(keycode_to_modifier(xc->xmodmap, keycode) != 0) 
	continue;

      keysym = XKeycodeToKeysym(self->display, keycode, 0);
      
      /* get a string from keysym or keycode */
      if(keysym != NoSymbol)
	keystr = XKeysymToString(keysym);
      else
	snprintf(keystr, 5, "0x%x", (int)keycode);

      /* abort on Ctrl-c */
      if (strcmp(keystr, "c") == 0 && 
	  get_modifiers(self->display) == ControlMask ) {
	
	XUngrabKeyboard(self->display, CurrentTime);
	return;
      }
      
      /* create a key object from the key string and add the currently
       * pressed modifiers */

      key = key_new(keystr);
      key->modifiers = get_modifiers(self->display);
      
      /* get the key specifier in Xchainkey format and print it */
      keyspec = key_to_str(key);

      printf("%s\n", keyspec);
      fflush(stdout);

      free(keyspec);
      free(key);
    }
  }
}

void xc_find_config(XChainKeys_t *self) {
  
  /* find the config file while respecting XDG_CONFIG_HOME */

  int n = 4096;
  self->config = (char *) calloc(n, sizeof(char));
  
  strncpy(self->config, getenv("HOME"), n);
  
  if(getenv("XDG_CONFIG_HOME") != NULL ) {
    strncpy(self->config, getenv("XDG_CONFIG_HOME"), n);
    strcat(self->config, "/xchainkeys/xchainkeys.conf");
  } 
  else {
    strcat(self->config, "/.config/xchainkeys/xchainkeys.conf");
  }
}

void xc_parse_config(XChainKeys_t *self) {

  FILE *f;
  char *buffer = (char *) calloc(4096, sizeof(char));
  char *argument= (char *) calloc(4096, sizeof(char));
  char *argument_ptr = argument;
  char *line, *token, *expect, *path;
  const char *ws = " \t"; 
  int linenum = 0;
  int len, pos;

  Key_t *key;
  Binding_t *binding;
  Binding_t *parent;
  Binding_t *existing;

  int feedback = True;

  char *font = calloc(512, sizeof(char));
  strcpy(font, "fixed");

  char *fg = calloc(64, sizeof(char));
  strcpy(fg, "black");

  char *bg = calloc(64, sizeof(char));
  strcpy(bg, "white");

  /* try to open config file */
  f = fopen(self->config, "r");
  
  if(f == NULL) {
    fprintf(stderr, "%s: error: '%s': %s\n", 
	    PACKAGE_NAME, self->config, strerror(errno));
    fflush(stderr);
    exit(EXIT_FAILURE);
  }

  if (self->debug) { 
    printf("Parsing config file %s\n", self->config);
    fflush(stdout);
  }

  /* parse file */
  while(fgets(buffer, 4096, f) != NULL) {
    linenum++;
    line = buffer;

    /* discard whitespace at the beginning */
    line += strspn(line, ws);
    
    /* ignore empty lines and comments */
    if( line[0] == '\n' || line[0] == '#' )
      continue;

    /* discard newline at end of string */
    if( line[strlen(line)-1] == '\n' )
      line[strlen(line)-1] = '\0';

    /* parse options */
    if( strncmp(line, "timeout", 7) == 0 ) {
      line += 7;
      line += strspn(line, ws);
      line[strcspn(line, ws)] = '\0';
      self->timeout = (unsigned int)atoi(line);
      continue;
    }

    if( strncmp(line, "feedback", 8) == 0 ) {
      line += 8;
      line += strspn(line, ws);
      line[strcspn(line, ws)] = '\0';

      if(strcmp(line, "off") == 0)
	feedback = False;
      if(strcmp(line, "on") == 0)
        feedback = True;

      continue;
    }

    if( strncmp(line, "delay", 5) == 0 ) {
      line += 5;
      line += strspn(line, ws);
      line[strcspn(line, ws)] = '\0';
      self->delay = (unsigned int)atoi(line);
      continue;
    }

    if( strncmp(line, "hold", 4) == 0 ) {
      line += 4;
      line += strspn(line, ws);
      line[strcspn(line, ws)] = '\0';
      self->hold = (unsigned int)atoi(line);
      continue;
    }

    if( strncmp(line, "position", 8) == 0 ) {
      line += 8;
      line += strspn(line, ws);
      self->position = (char *) calloc(strlen(line)+1, sizeof(char));
      self->position = strncpy(self->position, line, strlen(line));
      continue;
    }

    if( strncmp(line, "font", 4) == 0 ) {
      line += 4;
      line += strspn(line, ws);
      line[strcspn(line, ws)] = '\0';
      strncpy(font, line, 512);
      font[strlen(line)] = '\0';
      continue;
    }

    if( strncmp(line, "foreground", 10) == 0 ) {
      line += 10;
      line += strspn(line, ws);
      line[strcspn(line, ws)] = '\0';
      strncpy(fg, line, 64);
      fg[strlen(line)] = '\0';
      continue;
    }

    if( strncmp(line, "background", 10) == 0 ) {
      line += 10;
      line += strspn(line, ws);
      line[strcspn(line, ws)] = '\0';
      strncpy(bg, line, 64);
      bg[strlen(line)] = '\0';
      continue;
    }

    /* parse bindings */

    pos = 0;
    parent = xc->root;
    binding = NULL;
    expect = "key";
    argument[0] = '\0';

    while(strlen(line)) {
      len = strcspn(line, ws);
      
      token = calloc((len+1), sizeof(char));
      strncpy(token, line, len);

      line += strcspn(line, ws);
      line += strspn(line, ws);

      if( token[0] == ':' )
	expect = "action";

      if (strcmp(expect, "key") == 0) {
	if ((key = key_new(token)) != NULL) {

	  /* if this key is already bound in the parent binding,
	   * then make that binding the parent binding for the next key
	   */
	  existing = binding_get_child_by_key(parent, key);

	  if( existing != NULL) {
	    if(line[0] == ':' || line[0] == '\0') {
	      path = binding_to_path(existing);
	      fprintf(stderr, "%s: line %d: '%s': already bound, skipping...\n",
		      PACKAGE_NAME, linenum, path);
	      fflush(stderr);
	      free(path);
	      free(key);
	      free(token);
	      goto next_line;
	    }
	    else {
	      parent = binding_get_child_by_key(parent, key);
	      free(key);
	      goto next_token;
	    }
	  }
	  
	  /* avoid binding :abort, :escape or :group at toplevel */
	  if(parent == xc->root) {
	    if(strncmp(line, ":abort", 6) == 0 ||
	       strncmp(line, ":escape", 7) == 0 ||
	       strncmp(line, ":group", 6) == 0 ) {

	      fprintf(stderr, "%s: line %d: '%s': " 
		      "action is invalid outside of chain, skipping...\n",
		      PACKAGE_NAME, linenum, line); 
	      fflush(stderr);
	      free(key);
	      free(token);
	      goto next_line;
	    }
	  }

	  /* create a binding for this key */
	  binding = binding_new();
	  binding->key = key;

	  /* append the new binding to the current parent */
	  binding_append_child(parent, binding);

	  /* make this binding the parent for the next */
	  parent = binding;
	}
	else {
	  fprintf(stderr,
		  "%s: line %d: '%s': invalid keyspec, skipping...\n",
		  PACKAGE_NAME, linenum, token);
	  fflush(stderr);
	  free(token);
	  goto next_line;
	}
      }
      
      if (strcmp(expect, "action") == 0) {
	binding_set_action(binding, token);
	expect = "argument";
	goto next_token;
      }

      if (strcmp(expect, "argument") == 0) {
	if(strlen(argument))
	  strcat(argument, " ");
	strncat(argument, token, 4096-strlen(argument)-1);
	goto next_token;
      }

    next_token:
      pos++;
      free(token);
    }
    /* all tokens parsed */
    
    if (binding != NULL) {
      
      /* parse the name portion of the argument */
      if(argument[0] == '"') {
	argument += 1;
	if(strrchr(argument, '"') == NULL) {
	  fprintf(stderr, 
		  "%s: line %d: missing closing double quote "
		  "for action name, ignoring arguments...\n",
		  PACKAGE_NAME, linenum);
	  fflush(stderr);
	  goto next_line;
	}
        len = strcspn(argument, "\"");
	strncpy(binding->name, argument, 128);
	binding->name[len] = '\0';
	argument += len + 1;
	argument += strspn(argument, ws);
      }

      /* append the argument to the current binding */
      if (strlen(argument)) {
	strncpy(binding->argument, argument, 4096);
      }
      argument = argument_ptr;
    }
  next_line:
    ;
  }
  fclose(f);
  
  binding_parse_arguments(self->root);
  binding_create_default_bindings(self->root);

  /* initialize popup window */
  self->popup = popup_new(self->display, font, fg, bg, self->position);
  self->popup->enabled = feedback;

  /* display settings and chains on stdout */
  if (self->debug) {
    printf("\n");
    printf("timeout %d\n", self->timeout);

    if (feedback) { 
      printf("feedback on\n");
      printf("delay %d\n", self->delay);
      printf("hold %d\n", self->hold);
      printf("position %s\n", self->position);
      printf("font %s\n", font);
      printf("foreground %s\n", fg);
      printf("background %s\n\n", bg);
    }
    else {
      printf("feedback off\n\n");
    }
    
    binding_list(self->root);
    printf("\n");
    fflush(stdout);
  }
  free(buffer);
  free(argument);
  free(font);
  free(fg);
  free(bg);
}

void xc_grab_prefix_keys(XChainKeys_t *self) {
  /* grab top level keys individually */
  int i;
  for (i=0; i<self->root->num_children; i++) {
    key_grab(self->root->children[i]->key);    
  }
}

void xc_mainloop(XChainKeys_t *self) {
  Binding_t *binding;
  Binding_t *reentry;
  XEvent event;
  KeyCode keycode;
  struct timeval tv;
  fd_set in;
  int i;

  while(True) {
    xc_grab_prefix_keys(self);

    if (xc->popup->timeout > 0) {

        FD_ZERO(&in);
        FD_SET(xc->connection, &in);
        tv.tv_sec = xc->popup->timeout / 1000;
        tv.tv_usec = (xc->popup->timeout % 1000) * 1000;
        if (!select(xc->connection + 1, &in, 0, 0, &tv)) {
            popup_hide(xc->popup);
            xc->popup->timeout = 0;
        }
    }

    XNextEvent(self->display, &event);

    if (event.type == KeyPress) {
      keycode = ((XKeyPressedEvent*)&event)->keycode;
      
      for( i=0; i<self->root->num_children; i++ ) {
	binding = self->root->children[i];
	
	if (key_get_keycode(binding->key) == keycode) {
	  if(binding->key->modifiers == get_modifiers(xc->display)) {

	    popup_hide(xc->popup);
	    xc->popup->timeout = 0;
	    
	    binding_activate(binding);
	    break;
	  }
	}
      }
    }
  reentry:
    if(xc->reentry != NULL) {
      reentry = xc->reentry;
      xc->reentry = NULL;
      binding_activate(reentry);
      goto reentry;
    }

    if(xc->reload) {
      xc_reload(self);
    }
  }
}

void xc_reset(XChainKeys_t *self) {
  int i;
  Binding_t *prefix;

  for(i=0; i<self->root->num_children; i++) {
    prefix = self->root->children[i];

    key_ungrab(prefix->key);
    binding_free(prefix);
    self->root->children[i] = NULL;
  }
  self->root->num_children = 0;

  self->reentry = NULL;
  self->reload = False;

  popup_free(self->popup);
  self->popup = NULL;
}

void xc_reload(XChainKeys_t *self) {  
  xc_reset(self);
  xc_parse_config(self);
  xc_grab_prefix_keys(self);
}

void xc_parse_options(XChainKeys_t *self, int argc, char **argv) {

  struct option options[] = {
    { "debug", no_argument, NULL, 'd' },
    { "help", no_argument, NULL, 'h' },
    { "version", no_argument, NULL, 'v' },
    { "keys", no_argument, NULL, 'k' },
    { "file", no_argument, NULL, 'f' },
    { 0, 0, 0, 0 },
  };
  int option, option_index;

  while (1) {

    option = getopt_long(argc, argv, "dhvkf:", options, &option_index);
    
    switch (option) {

    case 'f':
      strncpy(self->config, optarg, 4096);
      self->config[strlen(optarg)] = '\0';
      break;      

    case 'k':
      xc_show_keys(self);
      exit(EXIT_SUCCESS);

    case 'd':
      self->debug = True;
      version();
      printf("\n"); fflush(stdout);
      break;
      
    case 'h':
      usage();
      exit(EXIT_SUCCESS);
      break;
      
    case 'v':
      version();
      exit(EXIT_SUCCESS);
      break;      

    case -1:
      break;

    default:
      usage();
      exit(EXIT_FAILURE);
      break;
    }
    if(option == -1)
      break;
  }  
}

int main(int argc, char **argv) {

  xc = xc_new();

  xc_parse_options(xc, argc, argv);
  xc_parse_config(xc);
  xc_mainloop(xc);

  exit(EXIT_SUCCESS);
}
